File: src/ipc/ai.ts
import { call } from "./client";
import type {
  AiProfile,
  AiProfileInput,
  RoleBindings,
  TestResult,
  Role,
  ApiProtocol,
} from "./types.gen";

export const aiApi = {
  listProfiles: () => call<AiProfile[]>("ai_list_profiles"),
  upsertProfile: (input: AiProfileInput) =>
    call<AiProfile>("ai_upsert_profile", { input }),
  deleteProfile: (id: string) => call<void>("ai_delete_profile", { id }),
  testProfile: (id: string) => call<TestResult>("ai_test_profile", { id }),
  listModels: (profileId: string) =>
    call<string[]>("ai_list_models", { profileId }),
  getRoleBindings: () => call<RoleBindings>("ai_get_role_bindings"),
  setRoleBinding: (role: Role, profileId: string, model: string) =>
    call<void>("ai_set_role_binding", { role, profileId, model, params: null }),
  clearRoleBinding: (role: Role) => call<void>("ai_clear_role_binding", { role }),
  cancelRequest: (streamId: string) =>
    call<void>("ai_cancel_request", { streamId }),
  debugChat: (prompt: string) => call<string>("ai_debug_chat", { prompt }),
};

export const AI_PRESETS: { label: string; baseUrl: string; protocol: ApiProtocol }[] = [
  { label: "LM Studio", baseUrl: "http://localhost:1234/v1", protocol: "openai" },
  { label: "MLXBar", baseUrl: "http://127.0.0.1:11435/v1", protocol: "openai" },
  { label: "Ollama", baseUrl: "http://localhost:11434/v1", protocol: "openai" },
  { label: "OpenAI", baseUrl: "https://api.openai.com/v1", protocol: "openai" },
  { label: "Anthropic", baseUrl: "https://api.anthropic.com/v1", protocol: "anthropic" },
  { label: "Custom (OpenAI-compatible)", baseUrl: "", protocol: "openai" },
  { label: "Custom (Anthropic-compatible)", baseUrl: "", protocol: "anthropic" },
];

File: src-tauri/src/commands/ai.rs
use crate::domain::ai::*;
use crate::error::{AppError, AppResult};
use crate::services::ai;
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub fn ai_list_profiles(state: State<'_, AppState>) -> AppResult<Vec<AiProfile>> {
    state.with_db(ai::profiles::list)
}

#[tauri::command]
pub fn ai_upsert_profile(state: State<'_, AppState>, input: AiProfileInput) -> AppResult<AiProfile> {
    state.with_db(|db| ai::profiles::upsert(db, input))
}

#[tauri::command]
pub fn ai_delete_profile(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.with_db(|db| ai::profiles::delete(db, &id))
}

#[tauri::command]
pub async fn ai_test_profile(state: State<'_, AppState>, id: String) -> AppResult<TestResult> {
    let path = state.app_db_path.clone();
    ai::test_profile(&path, &id).await
}

#[tauri::command]
pub fn ai_get_role_bindings(state: State<'_, AppState>) -> AppResult<RoleBindings> {
    state.with_db(ai::profiles::get_bindings)
}

#[tauri::command]
pub fn ai_set_role_binding(
    state: State<'_, AppState>,
    role: Role,
    profile_id: String,
    model: String,
    params: Option<serde_json::Value>,
) -> AppResult<()> {
    state.with_db(|db| {
        ai::profiles::set_binding(db, role, &profile_id, &model, params.unwrap_or(serde_json::json!({})))
    })
}

#[tauri::command]
pub fn ai_clear_role_binding(state: State<'_, AppState>, role: Role) -> AppResult<()> {
    state.with_db(|db| ai::profiles::clear_binding(db, role))
}

#[tauri::command]
pub fn ai_cancel_request(state: State<'_, AppState>, stream_id: String) -> AppResult<()> {
    if state.streams.cancel(&stream_id) {
        Ok(())
    } else {
        Err(AppError::new("STREAM_NOT_FOUND", "error.ai.streamNotFound", stream_id))
    }
}

/// Debug helper (docs/08 Phase 3: "check via the debug search-result view").
/// Streams a one-shot chat completion for the `chat` role and returns the
/// stream id; deltas arrive on `stream://*`.
#[tauri::command]
pub async fn ai_debug_chat(
    app: tauri::AppHandle,
    state: State<'_, AppState>,
    prompt: String,
) -> AppResult<String> {
    let resolved = state
        .with_db(|db| ai::profiles::resolve(db, Role::Chat))?
        .ok_or_else(|| AppError::new("AI_NOT_CONFIGURED", "error.ai.notConfigured", "no chat model set"))?;
    let (stream_id, token) = state.streams.start();
    let reg = state.streams.clone();
    let messages = serde_json::json!([{ "role": "user", "content": prompt }]);
    let sid = stream_id.clone();
    tauri::async_runtime::spawn(async move {
        ai::stream_chat(&app, &reg, resolved, messages, sid, token).await;
    });
    Ok(stream_id)
}
